memory/memhotplug.py

```diff
@@ -0,0 +1,250 @@
+                return True
+            return False
+
+    @staticmethod
+    def run_stress(stresstime):
+        mem_free = memory.meminfo.MemFree.mb / 2
+        cpu_count = int(multiprocessing.cpu_count()) / 2
+        process.run("stress --cpu %s --io 4 --vm 2 --vm-bytes %sM --timeout %ss" %
+                    (cpu_count, mem_free, stresstime), ignore_status=True, sudo=True, shell=True)
+
+    def test(self):
+        if os.path.exists("%s/auto_online_blocks" % mem_path):
+            if not self.__is_auto_online():
+                self.hotplug_all(self.blocks_hotpluggable)
+        if 'all' in self.tests:
+            tests = ['hotplug_loop',
+                     'hotplug_toggle',
+                     'hotplug_ratio',
+                     'dlpar_mem_hotplug',
+                     'hotplug_per_numa_node']
+        else:
+            tests = self.tests.split()
+
+        for method in tests:
+            self.log.info("\nTEST: %s\n", method)
+            self.__clear_dmesg()
+            run_test = 'self.%s()' % method
+            eval(run_test)
+            msg = self.__error_check()
+            if msg:
+                collect_dmesg()
+                self.log.error('Test: %s. ERROR Message: %s', run_test, msg)
+            self.log.info("\nEND: %s\n", method)
+
+    def hotplug_loop(self):
+        self.log.info("\nTEST: hotunplug and hotplug in a loop\n")
+        for _ in range(self.iteration):
+            self.log.info("\nhotunplug all memory\n")
+            self.hotunplug_all(self.blocks_hotpluggable)
+            self.run_stress(self.stresstime)
+            self.log.info("\nReclaim back memory\n")
+            self.hotplug_all(self.blocks_hotpluggable)
+
+    def hotplug_toggle(self):
+        self.log.info("\nTEST: Memory toggle\n")
+        for _ in range(self.iteration):
+            for block in self.blocks_hotpluggable:
+                offline(block)
+                self.log.info("memory%s block hotunplugged" % block)
+                self.run_stress(self.stresstime)
+                online(block)
+                self.log.info("memory%s block hotplugged" % block)
+
+    def hotplug_ratio(self):
+        for ratio in ('25', '50', '75', '100'):
+            target = 0
+            self.log.info("\nTEST : Hotunplug %s%% of memory\n" % ratio)
+            num = len(self.blocks_hotpluggable) * int(ratio)
+            if num % 100:
+                target = (num / 100) + 1
+            else:
+                target = num / 100
+            for block in self.blocks_hotpluggable:
+                if target > 0:
+                    offline(block)
+                    target -= 1
+                    self.log.info("memory%s block offline" % block)
+            self.run_stress(self.stresstime)
+            self.log.info("\nReclaim all memory\n")
+            self.hotplug_all(self.blocks_hotpluggable)
+
+    def dlpar_mem_hotplug(self):
+        if 'PowerNV' not in open('/proc/cpuinfo', 'r').read():
```

Comment:
i think this is specific powerpc can we have check for that ?

Reply:
ok